server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -106,6 +108,8 @@ private static void internalParseDocument(RootObjectMapper root, MetadataFieldMa
             parseObjectOrNested(context, root);
         }
 
+        PostParsePhase.executePostParsePhases(lookup, context);
+
         for (MetadataFieldMapper metadataMapper : metadataFieldsMappers) {
             metadataMapper.postParse(context);
```

Comment:
So I originally had a `postParse` method on Mapper that was just called recursively here; however, this doesn't play well with scripted fields that refer to other scripted fields.  In order to make sure scripts are executed in the right order, we need to know up front which fields have scripts on them, so that in the special index-time lookup we can intercept a lookup on an unexecuted script field and execute it.  And we only want to execute a script once, so we need something to hold this state, which ends up giving us this PostParsePhase.  Once we're collecting this set of fields up-front anyway, it made sense to remove the general postParse method.  I agree about the naming though, I'll see if I can come up with something more descriptive.

Reply:
it is not only about naming though, it's the complexity added by the executor concept, and the collection of all the executors. Couldn't we have another way to collect upfront all of the fields that declare a script to achieve what the post parse phase achieves?